service/storage_proxy.cc

```diff
@@ -5735,6 +5736,14 @@ storage_proxy::query_singular(lw_shared_ptr<query::read_command> cmd,
     exec.reserve(partition_ranges.size());
 
     schema_ptr schema = local_schema_registry().get(cmd->schema_version);
+    if (schema->is_view() && !schema->view_info()->base_info()) {
+        // If this is a reverse read, the reverse schema might have just been added
```

Comment:
The registry is supposed to provide valid schemas, and client code which looks them up shouldn't be responsible for "fixing" them.

So I think this should be done centrally, "when the reverse schema is added to the schema registry".

Reply:
I didn't mean to do it in the schema_registry, but at the site which populates the registry with the schema.

Which site is that?

Reply:
So at the time of get_or_load().

For example, we already do this in get_schema_definition():

```

    }).then([&storage_proxy] (schema_ptr s) {
        // If this is a view so this schema also needs a reference to the base
        // table.
        if (s->is_view()) {
            if (!s->view_info()->base_info()) {
                auto& db = storage_proxy.local_db();
                // This line might throw a no_such_column_family
                // It should be fine since if we tried to register a view for which
                // we don't know the base table, our registry is broken.
                schema_ptr base_schema = db.find_schema(s->view_info()->base_id());
                s->view_info()->set_base_info(s->view_info()->make_base_dependent_view_info(*base_schema));
            }
        }
        return s;
    });
```

and in global_schema_ptr.

So by that token, schema::get_reversed() should also do it. Maybe it's only about taking it from the original unreversed schema?

Reply:
btw, it looks like there is another bug related to views here. We have a potential issue in case schema_registry loses unfrozen schema_ptr (if all live schema_ptr die) and then recreates it from the frozen version. It should restore the base info as well, but it doesn't. We should cache the base info in addition to frozen schema.

Reply:
> btw, it looks like there is another bug related to views here. We have a potential issue in case schema_registry loses unfrozen schema_ptr (if all live schema_ptr die) and then recreates it from the frozen version. It should restore the base info as well, but it doesn't. We should cache the base info in addition to frozen schema.

I posted https://github.com/scylladb/scylladb/pull/21862 where I try to implement this, please take a look